### PR TITLE
feat: Publish Docker image to Github Registry instead of Docker Hub 

### PR DESCRIPTION
**Pull request by armenzg**
The overall changes in here are:

- Upload a [Docker image](https://github.com/orgs/getsentry/packages/container/package/action-release-image) to Github's registry
- Make this action use the Github registry image instead of the Docker Hub one

Fixes API-2791

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -7,7 +7,7 @@ RUN export YARN_CACHE_FOLDER="$(mktemp -d)" \
 COPY . .
 RUN yarn build
 
-FROM node:12-alpine
+FROM node:12-alpine as app
```
Comment (armenzg) [inline Dockerfile:10]: This may not be needed. I'm using `--target app` because I only want to tag this image and not the builder.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
+    # In the case of the Docker builder iamge, we should always see that it's using the remote cache
+    # XXX: Would a Docker push notice that is already there?
+    - name: Docker build images
+      run: |
+        docker buildx build . \
+          --build-arg BUILDKIT_INLINE_CACHE=1 \
+          -t ghcr.io/getsentry/action-release-image:latest \
+          -t ghcr.io/getsentry/action-release-image:${{ github.sha }} \
+          --cache-from ghcr.io/getsentry/action-release-image:latest \
+          --target app
```
Comment (armenzg) [inline .github/workflows/build.yml:36]: In a previous commit I was building both the builder and the app and publishing both, however, since the builder image is always being invalidated I will just build the app target.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -3,20 +3,75 @@ on:
   push:
     branches:
       - master
+      # XXX: Remove this before merging
+      - armenzg/ci/docker-release-to-ghr
```
Comment (armenzg) [inline .github/workflows/build.yml:7]: TODO: Remove after approval.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -3,20 +3,75 @@ on:
...
+        # are part of the release
+        with:
+          fetch-depth: 0
+
+      # Created "End to end action-release Integration" integration under Ecosystem's org
+      - name: Create Sentry release
+        # XXX: We need to test both master and v1 (official one)
+        # XXX: We need to create a v1 tag for the release to be completed
+        # XXX: Change this to master
+        uses: getsentry/action-release@armenzg/ci/docker-release-to-ghr
```
Comment (armenzg) [inline .github/workflows/build.yml:71]: TODO: Remove before merging

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -3,20 +3,75 @@ on:
...
+      - name: Checkout
+        uses: actions/checkout@v3
+        # This is necessary in order for the action to determine what commits
+        # are part of the release
+        with:
+          fetch-depth: 0
+
+      # Created "End to end action-release Integration" integration under Ecosystem's org
+      - name: Create Sentry release
+        # XXX: We need to test both master and v1 (official one)
```
Comment (armenzg) [inline .github/workflows/build.yml:68]: We will do this on another PR.

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -4,14 +4,16 @@ on:
   push:
     branches:
       - master
-      - 'releases/*'
```
Comment (armenzg) [inline .github/workflows/test.yml:7]: We have no branches named like this.

**Review by armenzg — COMMENTED** [review]
Here's a created release from this branch:
image

**Comment by armenzg** [comment]
Anyone that can please review this? Thanks!

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -1,22 +1,76 @@
...
+    # XXX: Remove comment
+    # needs: test
     runs-on: ubuntu-latest
     steps:
     - name: Checkout repo
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
       with:
+        # XXX: Why is this necessary?
         ref: ${{ github.head_ref }}
```
[the remark this answers is not in the document]
Comment (leeandher) [reply .github/workflows/build.yml:15]: Found this ([docs](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context))
> The `head_ref`  or source branch of the pull request in a workflow run. This property  is only available when the event that triggers a workflow run is either pull_request or pull_request_target.

So it is the merge commit :+1:

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -5,18 +5,70 @@ on:
...
       with:
-        username: ${{ secrets.DOCKER_USERNAME }}
-        password: ${{ secrets.DOCKER_PASSWORD }}
-        repository: sentryintegrations/sentry-github-action-release
-        tags: latest
+        environment: production
+
+    - name: Login and publish
+      run: |
+        echo ${{ secrets.GITHUB_TOKEN }} | docker login ghcr.io -u $GITHUB_ACTOR --password-stdin
```
Comment (leeandher) [inline .github/workflows/build.yml:45]: Is this `echo` statement leftover from debugging? Not sure it's meant to be here.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -5,18 +5,70 @@ on:
...
+        # The latest tag should only be published on `master`
+        docker push ghcr.io/getsentry/action-release-image:latest
+
+  # End to end testing: what the users would do
+  # XXX: TODO: Notify the team if this fails
+  test-action-release:
```
Comment (leeandher) [inline .github/workflows/build.yml:51]: Should this be it's own action so that we can test it before merging our PRs/community PRs as well?

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -5,18 +5,70 @@ on:
...
 
-    - name: Docker
-      uses: docker/build-push-action@v1
+    # BUILDKIT_INLINE_CACHE creates the image in such a way that you can
+    # then use --cache-from (think of a remote cache)
+    # This feature can be used because of using the buildx plugin
+    # In the case of the Docker builder iamge, we should always see that it's using the remote cache
+    # NOTE: Sadly, the COPY line from the builder image (in the Dockerfile) always invalidates the cache,
+    # thus, we do not benefit from using --cache-from. The multi-stage build still gives us a smaller footprint
+    # TODO: Optimize the Dockerfile to take advantage of proper caching
```
Comment (leeandher) [inline .github/workflows/build.yml:23]: Thanks for the explanation! It'll help a lot if anyone tries to debug/understand this later on 👏

File: .github/actions/use-local-dockerfile/action.yml [hunk 11]
```diff
@@ -1,13 +1,14 @@
...
     - name: Replace image on action.yml
       shell: bash
       env:
-        ESC_DOCKER_IMAGE_LINK: 'docker\:\/\/sentryintegrations\/sentry-github-action-release\:latest'
-      run: | 
-        mv action.yml previous.yml 
+        ESC_DOCKER_IMAGE_LINK: 'docker\:\/\/ghcr.io\/getsentry\/action-release-image\:latest'
+      run: |
+        mv action.yml previous.yml
         sed "s/$ESC_DOCKER_IMAGE_LINK/Dockerfile/" previous.yml >> action.yml
```
Comment (asottile-sentry) [inline .github/actions/use-local-dockerfile/action.yml:13]: fun little sed trick -- you can use any character as the delimiter so you don't need the complex escapes

so you can change the `IMAGE_LINK` to be the literal value and use `sed "s|$ESC_DOCKER_IMAGE_LINK|Dockerfile/"`

Comment (asottile-sentry) [reply .github/actions/use-local-dockerfile/action.yml:13]: this can also be `sed -i` and then you can remove the `mv` commands (note this requires gnu sed so won't work on macos, but that's fine here!)

**Review by leeandher — COMMENTED** [review]
Left a few comments/questions, only concern is the stray `echo` command. Looks good!

File: .github/workflows/build.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply .github/workflows/build.yml:15]: I would remove this and the line above -- github will by default check out the right thing

File: .github/workflows/build.yml (hunk 2 above)
Comment (asottile-sentry) [reply .github/workflows/build.yml:30]: oh hmm, you could probably push both, then pull both, then use two `--cache-from` arguments to make it not invalidated each time?

File: .github/workflows/build.yml (hunk 8 above)
Comment (asottile-sentry) [inline .github/workflows/build.yml:45]: neat little shell trick -- `docker login ... <<< ${{ secrets.GITHUB_TOKEN }}` -- https://youtu.be/Xf_82stIbB8

File: Dockerfile [hunk 12]
```diff
@@ -7,7 +7,7 @@ RUN export YARN_CACHE_FOLDER="$(mktemp -d)" \
 COPY . .
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply Dockerfile:7]: ah -- probably need a `.dockerignore` with `.git` in it? -- https://youtu.be/dEZ5WVWLris

**Review by asottile-sentry — APPROVED** [review]
a bunch of small things -- but looks good overall!

File: .github/workflows/build.yml (hunk 8 above)
Comment (asottile-sentry) [reply .github/workflows/build.yml:45]: it passes the password as stdin to `docker login` (which is expecting it there via `--password-stdin`) -- the idea there is to [avoid passwords in commands](https://youtu.be/SqXOchNKV_c)

File: .github/workflows/build.yml (hunk 9 above)
Comment (armenzg) [reply .github/workflows/build.yml:51]: Probably. Let me tackle this on a follow up PR

File: .dockerignore [hunk 13]
```diff
@@ -0,0 +1,5 @@
+# These files will be ignore by Docker when creating a build context
+# In other words, if a file should not be inside of the Docker container it should be
+# added to the list, otherwise, it will invalidate cache layers and have to rebuild
+# all layers after a COPY command
+.github
```
Comment (asottile-sentry) [inline .dockerignore:5]: ah the `.git` directory is the more important one here

File: .github/workflows/build.yml [hunk 14]
```diff
@@ -51,8 +51,11 @@ jobs:
     - name: Login and publish
       run: |
         docker login ghcr.io -u $GITHUB_ACTOR --password-stdin <<< ${{ secrets.GITHUB_TOKEN }}
-        docker push ghcr.io/getsentry/action-release-build-image:latest
-        docker push ghcr.io/getsentry/action-release-image:latest
+        docker push ghcr.io/getsentry/action-release-builder-image:latest || rm .docker/config.json
+        docker push ghcr.io/getsentry/action-release-image:latest || rm .docker/config.json
+        # WARNING! Your password will be stored unencrypted in /home/runner/.docker/config.json.
+        # Read https://docs.docker.com/engine/reference/commandline/login/#credentials-store
+        rm .docker/config.json
```
Comment (asottile-sentry) [inline .github/workflows/build.yml:58]: this will need to be `~/.docker/config.json`

**Review by armenzg — COMMENTED** [review]
I've removed the job that would test the action from master.

Something got messed up when it started trying to report the release from the pull_request.
I will look into it in the future.

**Comment by armenzg** [comment]
@leeandher @asottile-sentry I've addressed your concerns. I've reduced the scope by not testing the action and I've started publishing the builder Docker image. `--cache-from` is now working properly thanks to the `.dockerignore` change.

**Review by asottile-sentry — APPROVED** [review]
